### PR TITLE
Typegen support for the develop command

### PR DESCRIPTION
**Pull request by Convly**
### What does it do?

- It makes automatic type generation parts of the `develop` command core behavior. It cannot be disabled.
- Always clean up the "generated" types' folder before starting a type gen process
- It makes sure the "generated" types folder is removed before building the user project during the `develop` reload 

### Why is it needed?

With the introduction of the new Strapi's type system, we're improving the TS integration in our toolchain in order to provide a better developer experience. The automatic type generation is part of this & this PR aims to both fix existing issues related to generated types preventing the app to run and enable the automatic type generation for all users

Additional context: https://github.com/strapi/strapi/pull/17349#issuecomment-1639921212

### How to test it?

- `cd examples/kitchensink-ts`
- `yarn strapi ts:generate-types`
- `yarn develop`
- Create a new content type
- The server should reload successfully, and the types located in `types/generated/content-types.d.ts` should have been updated
- Try the same process with a javascript example app such as `examples/getstarted`

### Related issue(s)/PR(s)

fix [#16560 ](https://github.com/strapi/strapi/issues/17084)

### CODE REVIEW

File: packages/core/strapi/lib/commands/actions/develop/action.js [hunk 1]
```diff
@@ -47,6 +47,9 @@ module.exports = async ({ build, watchAdmin, polling, browser }) => {
 };
 
 const primaryProcess = async ({ distDir, appDir, build, isTSProject, watchAdmin, browser }) => {
+  // Make sure generated types are deleted before building or re-starting the application
+  tsUtils.cleanupGeneratedTypes({ pwd: appDir });
```
Comment (Convly) [inline packages/core/strapi/lib/commands/actions/develop/action.js:51]: Not sure if we should cleanup on reload only for TS project or for both JS & TS. Open to suggestions

Same applies for [L.89-90](https://github.com/strapi/strapi/pull/17349/files#diff-e98af4f85f1bf3c732c5f74eb70bd7805dd439c6e459b1ea7e9cde63d2c4ec5dR89-R90)

**Comment by joshuaellis** [comment]
How come we can't disable this? IMO I don't see a huge issue with it being opt-out by default, but feels odd we're enforcing it with no way to opt out...? 🤔 How does this effect startup / restart times etc?

**Comment by Convly** [comment]
> How come we can't disable this? IMO I don't see a huge issue with it being opt-out by default, but feels odd we're enforcing it with no way to opt out...? 

The goal here is to make type generation part of the overall DX experience as a core feature of the "reload" provided with the develop command. Allowing to opt out of this would be the same as trying to opt-out of the reload on file change (used on content type generation or code edition), it would break the flow and potentially break user applications built on TypeScript as the existing types would end up being out of sync with the new application structure. (see related issue in the description)

> How does this effect startup / restart times etc

I was aiming at making some experiments while the PR is in draft to confirm it won't have a huge impact.
Based on the tests I've done for the moment for an application with ~30 content types and ~7 components, it takes between 221ms and 276ms to generate everything.

**Comment by Convly** [comment]
FYI: This is on hold as we're gonna investigate more to try to find a solution that would work everywhere & every time

**Comment by Convly** [comment]
Notes: Another option would be to implement a custom compiler host that could remove/replace the source files associated to the generated types in some scenarios.

The main advantage is that it prevents any delete operation on the disk & instead simulate everything

For both solutions, if the shared registries are extended from somewhere else, it would still lead to errors

**Comment by Convly** [comment]
Closed by https://github.com/strapi/strapi/pull/17956

**Pull request closed without merging**